single-kernel/black_scholes.cpp

```diff
@@ -0,0 +1,267 @@
+    double candidate_answer;
+    int flag;
+    flag = (x < 0);
+    x = (x < 0) ? -x : x;
+    k = 1.0 / (1.0 + 0.2316419 * x);
+    accum = A4 + A5 * k;
+    accum = k * accum + A3;
+    accum = k * accum + A2;
+    accum = k * accum + A1;
+    accum = k * accum;
+    n = exp(-0.5 * x * x);
+    n *= INV_ROOT2PI;
+    candidate_answer = 1.0 - n * accum;
+    return (flag ? 1.0 - candidate_answer : candidate_answer);
+  }
+
+
+  double bsop_reference(int cpflag, double S0, double K, double r, double sigma, double T) {
+    double d1, d2, c, p, Nd1, Nd2, expval, answer;
+    d1 = log(S0 / K) + (r + 0.5 * sigma * sigma) * T;
+    d1 /= (sigma * sqrt(T));
+    expval = exp(-r * T);
+    d2 = d1 - sigma * sqrt(T);
+    Nd1 = N(d1);
+    Nd2 = N(d2);
+    c = S0 * Nd1 - K * expval * Nd2;
+    p = K * expval * (1.0 - Nd2) - S0 * (1.0 - Nd1);
+    answer = cpflag ? c : p;
+    return answer;
+  }
+
+public:
+  BlackScholes(const BenchmarkArgs& _args) : args(_args) {}
+
+  void setup() {
+    size = args.problem_size;     // input size defined by the user
+    local_size = args.local_size; // set local work_group size
+    // declare some variables for intializing data
+    int idx;
+    int S0Kdex, rdex, sigdex, Tdex;
+    FLOAT S0_array[4] = {42.0, 30.0, 54.0, 66.0};
+    FLOAT K_array[16] = {
+        40.0, 36.0, 44.0, 48.0, 24.0, 28.0, 32.0, 36.0, 48.0, 52.0, 56.0, 60.0, 60.0, 64.0, 68.0, 72.0};
+    FLOAT r_array[4] = {0.1, 0.09, 0.11, 0.12};
+    FLOAT sigma_array[4] = {0.2, 0.15, 0.25, 0.30};
+    FLOAT T_array[4] = {0.5, 0.25, 0.75, 1.0};
+    idx = 0;
+    /* Pointers used to allocate memory and split that memory into input and output arrays */
+    /* These pointers point to the data buffers needed for Black Scholes computation */
+    FIXED* cpflag;
+    FLOAT *S0, *K, *r, *sigma, *T, *answer;
+
+    cpflag = (FIXED*)malloc(size * sizeof(int));
+    S0 = (FLOAT*)malloc(size * sizeof(FLOAT));
+    K = (FLOAT*)malloc(size * sizeof(FLOAT));
+    r = (FLOAT*)malloc(size * sizeof(FLOAT));
+    sigma = (FLOAT*)malloc(size * sizeof(FLOAT));
+    T = (FLOAT*)malloc(size * sizeof(FLOAT));
+    answer = (FLOAT*)malloc(size * sizeof(FLOAT));
+
+    /* Here we load some values to simulate real-world options parameters.
+     * Users who wish to provide live data would replace this clause
+     * with their own initialization of the arrays. */
+    for(int k = 0; k < size; ++k) {
+      int* temp_int;
+      Tdex = (idx >> 1) & 0x3;
+      sigdex = (idx >> 3) & 0x3;
+      rdex = (idx >> 5) & 0x3;
+      S0Kdex = (idx >> 7) & 0xf;
+
+      temp_int = (int*)&cpflag[k];
+      temp_int[0] = (idx & 1) ? 0xffffffff : 0;
+      if(sizeof(FLOAT) == 8)
+        temp_int[1] = (idx & 1) ? 0xffffffff : 0;
+
+      S0[k] = S0_array[S0Kdex >> 2];
+      K[k] = K_array[S0Kdex];
+      r[k] = r_array[rdex];
+      sigma[k] = sigma_array[sigdex];
+      T[k] = T_array[Tdex];
+      answer[k] = 0.0f;
+      idx++;
+    }
+
+    // Il kernel è monodimensionale
```

Comment:
While I do, of course, greatly appreciate the beauty of the Italian language, I do think that the comments in SYCL-bench should be in English ;)